File: python/Advanced Graphs/min_cost_connect_points.py
from typing import List
import heapq

class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        """
        Find the minimum cost to connect all points using Manhattan distance.
        
        Approach: Prim's Algorithm for Minimum Spanning Tree (MST)
        1. Build adjacency list with Manhattan distances between all point pairs
        2. Use Prim's algorithm with min-heap to find MST
        3. Start from point 0, add edges with minimum cost until all points connected
        
        Time Complexity: O(n² log n)
        - Building adjacency list: O(n²) for all point pairs
        - Prim's algorithm: O(n² log n) in worst case
          - Each point can be processed at most n times
          - Each heap operation is O(log n)
        
        Space Complexity: O(n²)
        - Adjacency list: O(n²) for storing all edges
        - Min-heap: O(n²) in worst case
        - Visit set: O(n)
        """
        N = len(points)
        
        # Build adjacency list: {point_index: [[cost, neighbor_point], ...]}
        # Each point connects to all other points with Manhattan distance cost
        adj = {i: [] for i in range(N)}
        for i in range(N):
            x1, y1 = points[i]
            for j in range(i + 1, N):
                x2, y2 = points[j]
                # Manhattan distance: |x1-x2| + |y1-y2|
                dist = abs(x1 - x2) + abs(y1 - y2)
                adj[i].append([dist, j])
                adj[j].append([dist, i])
        
        # Prim's Algorithm Implementation
        res = 0  # Total minimum cost
        visit = set()  # Track visited points
        minH = [[0, 0]]  # Min-heap: [cost, point_index], start with point 0
        
        # Continue until all points are connected (MST has N-1 edges)
        while len(visit) < N:
            cost, i = heapq.heappop(minH)  # Get minimum cost edge
            
            # Skip if point already visited (avoid cycles)
            if i in visit:
                continue
                
            # Add cost to result and mark point as visited
            res += cost
            visit.add(i)
            
            # Add all unvisited neighbors to heap
            for neiCost, nei in adj[i]:
                if nei not in visit:
                    heapq.heappush(minH, [neiCost, nei])
        
        return res


# Test cases
def test_min_cost_connect_points():
    """Test cases for minCostConnectPoints function"""
    solution = Solution()
    
    # Test case 1: Simple 3-point triangle
    points1 = [[0, 0], [2, 2], [3, 10]]
    expected1 = 13  # Connect (0,0) to (2,2) = 4, then (2,2) to (3,10) = 9
    result1 = solution.minCostConnectPoints(points1)
    print(f"Test 1: {result1 == expected1}, Expected: {expected1}, Got: {result1}")
    
    # Test case 2: 4 points in square
    points2 = [[3, 12], [-2, 5], [-4, 1]]
    expected2 = 18  # Minimum spanning tree cost
    result2 = solution.minCostConnectPoints(points2)
    print(f"Test 2: {result2 == expected2}, Expected: {expected2}, Got: {result2}")
    
    # Test case 3: Single point
    points3 = [[0, 0]]
    expected3 = 0  # No connections needed
    result3 = solution.minCostConnectPoints(points3)
    print(f"Test 3: {result3 == expected3}, Expected: {expected3}, Got: {result3}")
    
    # Test case 4: Two points
    points4 = [[0, 0], [1, 1]]
    expected4 = 2  # Manhattan distance between points
    result4 = solution.minCostConnectPoints(points4)
    print(f"Test 4: {result4 == expected4}, Expected: {expected4}, Got: {result4}")
    
    # Test case 5: Points in line
    points5 = [[0, 0], [1, 0], [2, 0], [3, 0]]
    expected5 = 3  # Connect consecutive points: 1+1+1 = 3
    result5 = solution.minCostConnectPoints(points5)
    print(f"Test 5: {result5 == expected5}, Expected: {expected5}, Got: {result5}")


if __name__ == "__main__":
    test_min_cost_connect_points()

File: C++/Advanced Graphs/min_cost_connect_points.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        /*
        Approach:
        - Treat each point as a node in a complete graph.
        - Edges are the Manhattan distances between points.
        - Use Prim’s Algorithm to find the Minimum Spanning Tree (MST) of this graph.

        Time Complexity: O(n² log n)
            - Building adjacency list: O(n²)
            - Each heap operation: O(log n), up to O(n²) operations.

        Space Complexity: O(n²)
            - Adjacency list: O(n²)
            - Priority queue: O(n²)
            - Visited set: O(n)
        */

        int N = points.size();

        // Build adjacency list: for each point, store all edges (cost, neighbor)
        unordered_map<int, vector<pair<int, int>>> adj;

        for (int i = 0; i < N; ++i) {
            int x1 = points[i][0], y1 = points[i][1];
            for (int j = i + 1; j < N; ++j) {
                int x2 = points[j][0], y2 = points[j][1];
                int dist = abs(x1 - x2) + abs(y1 - y2);  // Manhattan distance
                adj[i].push_back(make_pair(dist, j));
                adj[j].push_back(make_pair(dist, i));
            }
        }

        int res = 0;                   // Total minimum cost
        unordered_set<int> visited;   // Visited points
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minH;

        // Start from point 0
        minH.push(make_pair(0, 0));  // (cost, point)

        while (visited.size() < N) {
            pair<int, int> top = minH.top();
            minH.pop();

            int cost = top.first;
            int i = top.second;

            if (visited.count(i)) continue;

            res += cost;
            visited.insert(i);

            for (size_t k = 0; k < adj[i].size(); ++k) {
                int neiCost = adj[i][k].first;
                int nei = adj[i][k].second;
                if (!visited.count(nei)) {
                    minH.push(make_pair(neiCost, nei));
                }
            }
        }

        return res;
    }
};


/*
================ LINE BY LINE SYNTAX EXPLANATION =====================

#include <bits/stdc++.h>
- Includes all standard C++ headers (vector, set, map, queue, etc.)

using namespace std;
- Avoids writing std:: before STL names.

class Solution
- Defines the Solution class, as required by LeetCode.

int minCostConnectPoints(vector<vector<int>>& points)
- Function that takes a vector of points and returns the minimum cost.

int N = points.size();
- Number of points.

unordered_map<int, vector<pair<int, int>>> adj;
- Adjacency list to store all edges from each point: point → [(cost, neighbor), …].

for (int i = 0; i < N; ++i)
- Outer loop to process each point.

int x1 = points[i][0], y1 = points[i][1];
- Get the x and y coordinates of point i.

for (int j = i + 1; j < N; ++j)
- Inner loop to process every other point after i.

int x2 = points[j][0], y2 = points[j][1];
- Get x and y of point j.

int dist = abs(x1 - x2) + abs(y1 - y2);
- Compute Manhattan distance.

adj[i].push_back(make_pair(dist, j));
adj[j].push_back(make_pair(dist, i));
- Add both edges (since the graph is undirected).

int res = 0;
- Variable to store the final minimum cost.

unordered_set<int> visited;
- Set of visited points to avoid cycles.

priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>>
- Min-heap (priority queue) of (cost, point) pairs.

minH.push(make_pair(0, 0));
- Start from point 0 with cost 0.

while (visited.size() < N)
- Keep running until all points are visited.

pair<int, int> top = minH.top();
minH.pop();
- Extract the minimum cost edge.

int cost = top.first;
int i = top.second;
- Extract cost and point from the pair.

if (visited.count(i)) continue;
- Skip if already visited.

res += cost;
visited.insert(i);
- Add point to MST and update total cost.

for (size_t k = 0; k < adj[i].size(); ++k)
- Iterate over all neighbors of point i.

int neiCost = adj[i][k].first;
int nei = adj[i][k].second;
- Extract neighbor cost and index.

if (!visited.count(nei))
- If neighbor is not yet visited.

minH.push(make_pair(neiCost, nei));
- Push neighbor edge into heap.

return res;
- Return the minimum cost to connect all points.

=======================================================================
*/
